lib/sqlitex/statement.ex

```diff
@@ -0,0 +1,185 @@
+defmodule Sqlitex.Statement do
+  @moduledoc """
+  Provides an interface for working with sqlite prepared statements.
+
+  ```
+  iex(2)> {:ok, db} = Sqlitex.open(":memory:")
+  iex(3)> Sqlitex.query(db, "CREATE TABLE data (id, name);")
+  []
+  iex(6)> {:ok, statement} = Sqlitex.Statement.prepare(db, "INSERT INTO data VALUES (?, ?);")
+  iex(7)> Sqlitex.Statement.bind_values(statement, [1, "hello"])
+  iex(8)> Sqlitex.Statement.fetch_all(statement)
+  []
+  iex(9)> {:ok, statement} = Sqlitex.Statement.prepare(db, "SELECT * FROM data;")
+  iex(10)> Sqlitex.Statement.fetch_all(statement)
+  [[id: 1, name: "hello"]]
+  iex(11)> Sqlitex.close(db)
+  :ok
+
+  ```
+  """
+
+  use Pipe
+
+  defstruct database: nil,
+            statement: nil,
+            column_names: [],
+            column_types: []
+
+  @doc """
+  Prepare a Sqlitex.Statement
+
+  ## Parameters
+
+  * `db` - The database to prepare the statement for.
+  * `sql` - The SQL of the statement to prepare.
+
+  ## Returns
+
+  * `{:ok, statement}` on success
+  * See `:esqlite3.prepare` for errors.
+  """
+  def prepare(db, sql) do
+    pipe_with &pipe_ok/2,
+      do_prepare(db, sql)
+      |> get_column_names
+      |> get_column_types
+  end
+
+  @doc """
+  Same as `prepare/2` but raises an error on error.
+
+  Returns a new statement otherwise.
+  """
+  def prepare!(db, sql) do
+    {:ok, statement} = prepare(db, sql)
+    statement
+  end
+
+  @doc """
+  Binds values to a Sqlitex.Statement
+
+  ## Parameters
+
+  * `statement` - The statement to bind values into.
+  * `values` - A list of values to bind into the statement.
+
+  ## Returns
+
+  * `{:ok, statement}` on success
+  * See `:esqlite3.prepare` for errors.
+
+  ## Value transformations
+
+  Some values will be transformed before insertion into the database.
+
+  * `nil` - Converted to :undefined
+  * `true` - Converted to 1
+  * `false` - Converted to 0
+  * `datetime` - Converted into a string.  See datetime_to_string
+  * `%Decimal` -  Converted into a number.
+  """
+  def bind_values(statement, values) do
+    case :esqlite3.bind(statement.statement, translate_bindings(values)) do
+      {:error, _}=error -> error
+      :ok -> {:ok, statement}
+    end
+  end
+
+  @doc """
+  Same as `bind_values/2` but raises an error on error.
+
+  Returns the statement otherwise.
+  """
+  def bind_values!(statement, values) do
```

Comment:
Though having said that, [the sqlite docs](https://www.sqlite.org/threadsafe.html) seem to suggest that "Serialized" mode is the default, and 

> In serialized mode, SQLite can be safely used by multiple threads with no restriction.

I'm not sure if that means you can bind a statement twice and avoid overwriting the original bindings though - I'll maybe add a test to see.